### ISSUE TITLE
Resource Server handles insufficient scope

### ISSUE THREAD

**Issue by jzheaux** [spring-projects/spring-security#5227]
### Summary

The OAuth2 Bearer Token error response includes the property and error code `scope` and `insufficient_scope` respectively. [1]

The initial validation of the bearer token in Spring Security is done in the authentication portion of the filter chain. Once this has happened, the filter chain sets an `Authentication` object representative of the valid bearer token. Problems in this phase of the request throw an `AuthenticationException`. Currently, there is a subclass of `AuthenticationException` that includes OAuth2 context that is used to correctly formulate the error response.

This first phase, however, doesn't take into account scopes. 

In the next phase, the `FilterSecurityInterceptor` steps in and checks the `Authentication` object for sufficient scope, as previously derived from the original bearer token. Problems in this phase of the request throw an `AccessDeniedException`.

To support returning `insufficient_scope` as an error code, `FilterSecurityInterceptor` and `MethodSecurityInterceptor` should be able to return enough context in an `AccessDeniedException` subclass to at least indicate that the error code is warranted and at best indicate which scopes are required in order for the request to be granted.

[1] https://tools.ietf.org/html/rfc6750#section-3.1

Related to #4887

**Comment by jgrandja** [comment]
Related #5132

**Comment by jzheaux** [comment]
Here are some challenges with this task:

1. Support for OAuth2 scopes today is through SpEL. This means we need to take action while in the context of expression evaluation or we need another way to express scopes.

2. The access denied code path isn't prepared for a downstream voter or other advice to propagate its reasons for a deny vote back out to the client. While such is doable, it probably can't be done within the context of this task.

In `spring-security-oauth` (i.e. legacy oauth), an `AccessDeniedException` is thrown from within the `OAuth2SecurityExpressionMethods`, and it may be possible for us to do the same with a custom `AccessDeniedException` and custom `AccessDeniedHandler`:

```java
public class OAuth2ExpressionsBean {
    public boolean hasScope(Authentication auth, String scope) {
        // ... if fails verification
        throw new InsufficientScopeException(scope);
    }
}
```
The limitations center around preemptively assuming that a missing scope means denial. For example:

```java
@PreAuthorize("@oauth2.hasScope(authentication, 'resource.read') or @otherAuthorities")
```
implies that the request isn't denied just because a scope is missing.

Though this can be overcome fairly easily with a custom bean:

```java
@PreAuthorize("@authorizer.authorize(authentication)")
```

@rwinch, @jgrandja any thoughts regarding implementation?

**Comment by jgrandja** [comment]
Closed via 40ccdb9
